### PR TITLE
feat: adding momento-redis compatibility client for go-lang

### PR DESCRIPTION
**Pull request by pratik151192**
This PR contains implementation for all the Momento scalar APIs (GET, SET, EXPIRE, TTL, DEL, SETNX). 
It has all the changes needed without the CI workflows for this project. If you want an incremental view, please look at https://github.com/momentohq/momento-go-redis-client/pull/2 where the interfaces were exposed without implementing anything.

The reason this PR directly tracks main is to save rebase nightmare for me as I have made a lot edits since that PR on common files!

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+name: Build
```
Comment (pratik151192) [inline .github/workflows/build.yml:1]: can be ignored as i'm working CIs

File: momento-redis/momento_redis.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package momento_redis
```
Comment (pratik151192) [inline momento-redis/momento_redis.go:1]: file that matters

File: momento-redis/marshaller.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package momento_redis
```
Comment (pratik151192) [inline momento-redis/marshaller.go:1]: file that matters

File: momento-redis/test_helpers/shared_context.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package helpers
```
Comment (pratik151192) [inline momento-redis/test_helpers/shared_context.go:1]: file that matters

File: momento-redis/scalar.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
+package momento_redis
```
Comment (pratik151192) [inline momento-redis/scalar.go:1]: file that matters

File: momento-redis/scalar_test.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
+package momento_redis_test
```
Comment (pratik151192) [inline momento-redis/scalar_test.go:1]: file that matters

**Comment by pratik151192** [comment]
This PR https://github.com/momentohq/momento-go-redis-client/pull/3 contains everything including the changes in this PR. The existing PR won't be merged and #3 will be pushed which contains everything.. but still keeping it around for an incremental view

File: MakeFile
Comment (pgautier404) [inline MakeFile:1]: Please rename to `Makefile`

File: momento-redis/scalar_test.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+	. "github.com/momentohq/momento-go-redis-client/momento-redis/test_helpers"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/redis/go-redis/v9"
+)
+
+var _ = Describe("Scalar methods", func() {
+	sContext := NewSharedContext()
+	BeforeEach(func() {
+		cacheName := "default_cache"
```
Comment (pgautier404) [inline momento-redis/scalar_test.go]: You probably want to use a UUID here to prevent collisions if tests are run in parallel., e.g.: 

`fmt.Sprintf("golang-redis-%s", uuid.NewString())`

File: README.md [hunk 8]
```diff
@@ -1,2 +1,190 @@
...
+You can use Momento as your cache engine for any Go project that support a redis-backed cache.
+
+## Usage
+
+To switch your existing `go-redis` application to use Momento, you only need to change the code where you construct your client object:
+
+<table>
+<tr>
+ <td width="50%">With go-redis client</td>
+ <td width="50%">With Momento's go-redis compatibility client</td>
```
Comment (pgautier404) [inline README.md]: This layout is showing the examples in a scrolling window that makes it hard to compare them side-by-side. I think it might work better to just have them one after another vertically on the page.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,192 @@
...
+	"github.com/momentohq/client-sdk-go/auth"
+	"github.com/momentohq/client-sdk-go/config"
+	"github.com/momentohq/client-sdk-go/momento"
+	"github.com/momentohq/momento-go-redis-client/momento-redis"
+)
+
+credential, _ := auth.NewEnvMomentoTokenProvider("MOMENTO_AUTH_TOKEN")
+cacheClient, _ := momento.NewCacheClient(config.LaptopLatest(), credential, 60*time.Second)
+// create cache; it resumes execution normally incase the cache already exists and isn't exceptional
+cacheClient.CreateCache(context.Background(), &momento.CreateCacheRequest {CacheName : "default_cache"})
```
Comment (pgautier404) [inline README.md]: Please go ahead and add error handling to the example here and the one near the bottom of this file.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,192 @@
...
+with ```-host x.x.x.x``` and ```-port xxxx```.
+
+You can run the example via `go run`.
+
+Here's an example run against Momento Cache:
+
+```bash
+cd examples/basic
+export MOMENTO_AUTH_TOKEN=<your momento auth token goes here>
+go run main.go -cacheName cache -authToken $MOMENTO_AUTH_TOKEN
```
Comment (pgautier404) [inline README.md]: I don't think it's a good idea from a security perspective to have users pass an auth token on the command line. Most of our SDK examples read the environment variable directly, and you should probably do that in this example as well.

File: examples/basic/main.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	"context"
+	"flag"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/momentohq/client-sdk-go/auth"
+	"github.com/momentohq/client-sdk-go/config"
+	"github.com/momentohq/client-sdk-go/momento"
+	momento_redis "github.com/momentohq/momento-go-redis-client/momento-redis"
```
Comment (pgautier404) [inline examples/basic/main.go]: Super nit-picky, but golang doesn't like snake case. Please rename to `momentoredis` or `momentoRedis`.

File: examples/basic/main.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+		client = initMomentoRedisClient(*o)
+		fmt.Println("\nUsing Momento as a backend for go-redis with cache name \"" + o.cacheName + "\"")
+
+	default:
+		panic("nothing found")
+	}
+	fmt.Println("\n-------------------------------------------------")
+
+	key := "Momento"
+	sResp := client.Set(context.Background(), key, "cache", 60*time.Second)
```
Comment (pgautier404) [inline examples/basic/main.go]: Instead of constructing a new context for each method call, we usually define a single `ctx` variable at the top of `main()` to pass around instead. We're not really using context at this point, but if we ever do we'll need it to be a persistent object.

File: momento-redis/momento_redis.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+func NewMomentoRedisClient(cacheClient momento.CacheClient, cacheName string) *MomentoRedisClient {
+	client := MomentoRedisClient{
+		cacheName: cacheName,
+		client:    cacheClient,
+	}
+	return &client
+}
+
+func (c *MomentoRedisClient) String() string {
```
Comment (pgautier404) [inline momento-redis/momento_redis.go]: Please rename this receiver to `m` to match the other receivers of this type in the `momento_redis` package. My IDE called this one out . . . I would have never noticed that on my own :-)

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+Thank you for taking your time to contribute to our Momento `go-redis` wrapper!
+<br/>
+This guide will provide you information to start your own development and testing.
+<br/>
+Happy coding :dancer:
+<br/>
+
+## Requirements :coffee:
+
+- Go 1.2 or above is required https://go.dev/doc/install
```
Comment (pgautier404) [inline CONTRIBUTING.md]: According to the `CONTRIBUTING` doc in the SDK, our minimum supported Go version is currently 1.18.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+<br/>
+
+## Tests :zap:
+
+### Run integration tests against Momento 
+
+```bash
+TEST_AUTH_TOKEN=<YOUR_AUTH_TOKEN>
+make test-momento
```
Comment (pgautier404) [inline CONTRIBUTING.md]:
```suggestion
TEST_AUTH_TOKEN=<YOUR_AUTH_TOKEN> make test-momento
```

or

```
export TEST_AUTH_TOKEN=<YOUR_AUTH_TOKEN>
make test-momento
```

Comment (pratik151192) [reply CONTRIBUTING.md]: it actually works on a single line as well! do you mean more for readability? Also, ```export``` isn't necessarily required

File: momento-redis/scalar_test.go [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+			sContext.Client = redis.NewClient(&redis.Options{
+				Addr: host + ":" + port,
+			})
+		case false:
+			credential, _ := auth.NewEnvMomentoTokenProvider(AuthTokenEnvVariable)
+			mClient, _ := momento.NewCacheClient(config.LaptopLatest(), credential, 60*time.Second)
+			sContext.MomentoClient = mClient
+			// create cache; it resumes execution normally incase the cache already exists and isn't exceptional
+			sContext.CreateCache(sContext.Ctx, mClient, cacheName)
+			sContext.Client = momentoredis.NewMomentoRedisClient(mClient, cacheName)
```
Comment (pgautier404) [inline momento-redis/scalar_test.go]: Can you go ahead and add error handling to these setup steps please? I had a problem with my auth token environment variable but the only error being surfaced in my test output was about `invalid memory address or nil pointer dereference`.

File: CONTRIBUTING.md (hunk 15 above)
Comment (pgautier404) [reply CONTRIBUTING.md]: If I just issue those commands as you have them there, one after another without using `export`, the tests fail with an `invalid memory address or nil pointer dereference` when it tries to use the credential provider.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+```
+### Run integration tests against Redis
+
+First run Redis either natively, run Redis in a Docker container, or do your development in a devcontainer. Here is an example of running Redis in a Docker container:
+
+```
+docker run -it -p 6379:6379 redis
+```
+
+Then run the tests against .NET6 and .NET Framework:
```
Comment (pgautier404) [inline CONTRIBUTING.md]: Probably want to update this :-)

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+### Run integration tests against Momento 
+
+```bash
+TEST_AUTH_TOKEN=<YOUR_AUTH_TOKEN>
+make test-momento
+```
+### Run integration tests against Redis
+
+First run Redis either natively, run Redis in a Docker container, or do your development in a devcontainer. Here is an example of running Redis in a Docker container:
```
Comment (pgautier404) [inline CONTRIBUTING.md:43]: I have no idea what a devcontainer is :-) I see you have a `.devcontainer` directory in the repo. Can you add some docs about how that works or a link to some docs? Honestly, the docker command is so dead simple to get someone up and running for testing, I wonder if it's really necessary?

Comment (pratik151192) [reply CONTRIBUTING.md:43]: you're right. removed!

File: CONTRIBUTING.md (hunk 17 above)
Comment (pratik151192) [reply CONTRIBUTING.md]: ah good catch!

File: CONTRIBUTING.md (hunk 15 above)
Comment (pratik151192) [reply CONTRIBUTING.md]: oh interesting! changed

File: README.md [hunk 19]
```diff
@@ -1,2 +1,218 @@
...
+You can use Momento as your cache engine for any Go project that support a redis-backed cache.
+
+## Usage
+
+To switch your existing `go-redis` application to use Momento, you only need to change the code where you construct your client object:
+
+<table>
+<tr>
+ <td width="50%">With go-redis client</td>
+ <td width="50%" valign="top">
```
Comment (pgautier404) [inline README.md]: Sorry to keep harping on this, but can you just pull it out of a table completely? When I look at https://github.com/momentohq/momento-go-redis-client/tree/final-impl-go-redis-client, I still have to scroll horizontally to see the code and it's impossible to see it all without some portion of it being cut off.

File: README.md [hunk 20]
```diff
@@ -1,2 +1,218 @@
...
+	credential, eErr := auth.NewEnvMomentoTokenProvider("MOMENTO_AUTH_TOKEN")
+	if eErr != nil {
+		panic("Failed to initialize credentials through auth token " + eErr.Error())
+	}
+	cacheClient, cErr := momento.NewCacheClient(config.LaptopLatest(), credential, 60*time.Second)
+	if cErr != nil {
+		panic("Failed to initialize Momento cache client " + cErr.Error())
+	}
+	// create cache; it resumes execution normally incase the cache already exists and isn't exceptional
+	cacheClient.CreateCache(context.Background(), &momento.CreateCacheRequest{CacheName: "default_cache"})
```
Comment (pgautier404) [inline README.md]: Please add error handling here in case there is a problem creating the cache.

File: examples/basic/main.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+var options interface{}
+var momentoCacheClient momento.CacheClient
+var momentoCacheName string
+
+func main() {
+	parseFlags()
+
+	// change this to the type momentoredis.MomentoRedisCmdable for compile-time checking
+	// var client momentoredis.MomentoRedisCmdable
```
Comment (pgautier404) [inline examples/basic/main.go]: Nice.

File: examples/basic/main.go [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+			" variable MOMENTO_AUTH_TOKEN?\n" + eErr.Error())
+	}
+	cacheClient, cErr := momento.NewCacheClient(config.LaptopLatest(), credential, time.Duration(options.defaultTTlSeconds)*time.Second)
+	if cErr != nil {
+		panic("Failed to initialize Momento cache client\n" + cErr.Error())
+	}
+	// create cache; it resumes execution normally incase the cache already exists and isn't exceptional
+	cacheClient.CreateCache(context.Background(), &momento.CreateCacheRequest{
+		CacheName: options.cacheName,
+	})
```
Comment (pgautier404) [inline examples/basic/main.go]: Please add error checking/handling here.

File: examples/basic/main.go [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		options = &RedisOptions{
+			host: *host,
+			port: *port,
+		}
+	} else {
+
+		if *cacheName == "" {
+			panic("Running in Momento mode: Momento cacheName (-cacheName) should be provided through command line arguments." +
+				"For Redis more, use flag -useRedis along with -host and -port")
```
Comment (pgautier404) [inline examples/basic/main.go]:
```suggestion
				" For Redis mode, use flag -useRedis along with -host and -port")
```

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+
+Then run the tests
+
+```
+make test-redis
+```
+
+This assumes the Redis server is running on `localhost:6379`.
+
+By running Redis on the local host, you can use the `redis-cli` to inspect the state of the Redis server as well as interactively debug the tests. We have also included a devcontainer config which will mount the entire coding environment into a Docker container, including the Redis server. See `.devcontainer` for more details.
```
Comment (pgautier404) [inline CONTRIBUTING.md]: Need to update this now that we've removed devcontainer.

Comment (pratik151192) [reply CONTRIBUTING.md]: ugh my bad

File: examples/README.md [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+#### Running examples with compile-time checking and Momento flavor interface
+
+If you'd like compile-time checking to tell you if you are using any APIs that we don't yet
+support, we provide our own `MomentoRedisCmdable` interface, which is a fully compatible subset of the official `go-redis`
+interface `Cmdable`, but explicitly lists out the APIs that we currently support.
+
+If you want the examples to be run using the Momento flavor interface, 
+change line number #37 on main.go and comment line number #38. With this one line of code change,
```
Comment (pgautier404) [inline examples/README.md]: I'd be careful about calling out specific line numbers just from a doc maintenance standpont. Something like "change the type from `redis.Cmdable` to `momentoredis.MomentoRedisCmdable` where the `client` variable is declared near the top of `main()`" ought to be explanatory enough but also stays correct if we add code that makes the line numbers incorrect.

File: momento-redis/scalar_test.go [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+				panic("Failed to initialize credentials through auth token. Did you export the environment" +
+					" variable TEST_AUTH_TOKEN?\n" + eErr.Error())
+			}
+			mClient, cErr := momento.NewCacheClient(config.LaptopLatest(), credential, 60*time.Second)
+			if cErr != nil {
+				panic("Failed to initialize Momento cache client\n" + cErr.Error())
+			}
+			sContext.MomentoClient = mClient
+			// create cache; it resumes execution normally incase the cache already exists and isn't exceptional
+			sContext.CreateCache(sContext.Ctx, mClient, cacheName)
```
Comment (pgautier404) [inline momento-redis/scalar_test.go]: Need error checking here please.

File: examples/basic/main.go (hunk 23 above)
Comment (pgautier404) [reply examples/basic/main.go]: Can you please add a space at the beginning of the "For redis mode..." string? It's being concatenated with the "Running in momento mode..." string but is missing a space between the sentences.

File: momento-redis/scalar_test.go [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+			if createErr != nil {
+				panic("Failed to create cache with cache name " + cacheName + "\n" + createErr.Error())
+			}
+			sContext.Client = momentoredis.NewMomentoRedisClient(mClient, cacheName)
+		}
+		DeferCleanup(func() {
+			if !sContext.UseRedis {
+				_, deleteErr := sContext.DeleteCache(sContext.Ctx, sContext.MomentoClient, cacheName)
+				if deleteErr != nil {
+					panic("Failed to create cache with cache name " + cacheName + "\n" + deleteErr.Error())
```
Comment (pgautier404) [inline momento-redis/scalar_test.go]:
```suggestion
					panic("Failed to delete cache with cache name " + cacheName + "\n" + deleteErr.Error())
```

**Review by pgautier404 — APPROVED** [review]
Thank you, this looks great!
